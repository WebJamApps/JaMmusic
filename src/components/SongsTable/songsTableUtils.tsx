
import superagent from 'superagent';
import type { ISong } from 'src/providers/Data.provider';

const editSong = (data: ISong, makeNewEditor:any, editor:any, picsForm:HTMLElement | null): boolean => {
  const songData = { ...data, modify: undefined };
  makeNewEditor({ ...editor, song: songData }); // this is storing the song globally in redux
  if (picsForm && picsForm.scrollIntoView) picsForm.scrollIntoView({ behavior: 'smooth' });
  return true;
};

const deleteSong = async (id: string, token:string): Promise<string> => { // eslint-disable-next-line no-restricted-globals
  const result = confirm('Deleting song, are you sure?');// eslint-disable-line no-alert
  if (result) {
    let res: superagent.Response;
    try {
      res = await superagent.delete(`${process.env.BackendUrl}/song/${id}`)
        .set('Authorization', `Bearer ${token}`).set('Accept', 'application/json');
    } catch (e) { return (e as Error).message; }
    if (res.status === 200) { window.location.reload(); return 'deleted pic'; }
    return `${res.status} ${res.body}`;
  }
  return 'no delete';
};

const makeSongButtons = (
  deleteSongId: string,
  editSongId: string,
  token: string,
  makeNewEditor:any,
  editor:any,
  song: ISong,
):JSX.Element => {
  const picsForm = document.getElementById('picsForm');
  return (
    <div>
      <button type="button" id={deleteSongId} onClick={() => deleteSong(song._id as string, token)}>Delete</button>
      <p>{' '}</p>
      <button
        type="button"
        id={editSongId}
        onClick={() => {
          editSong(song, makeNewEditor, editor, picsForm); // Here is what we need to change!
        }}
      >
        Edit
      </button>
    </div>
  );
};

const addButtons = (arr: ISong[], token:string, setNewEditor:any, editor:any): ISong[] => {
  const newArr = arr;/* eslint-disable security/detect-object-injection */
  for (let i = 0; i < arr.length; i += 1) {
    const deleteSongId = `deleteSong${newArr[i]._id}`;
    const editSongId = `editSong${newArr[i]._id}`;
    newArr[i].modify = makeSongButtons(deleteSongId, editSongId, token, setNewEditor, editor, newArr[i]);
  }
  return newArr;
};

export default { addButtons, editSong, deleteSong };
