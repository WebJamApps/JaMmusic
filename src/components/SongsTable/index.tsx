import React from 'react';
import MUIDataTable from 'mui-datatables';
import type { Dispatch, AnyAction } from 'redux';
import { ISong, SongsContext } from '../../providers/Songs.provider';
import songsTableUtils from './songsTableUtils';

type Props = {
  sData?:ISong[], token:string, dispatch: Dispatch<AnyAction>
};
const SongsTable:React.FC<Props> = ({ sData, token, dispatch }): JSX.Element => {
  const { test, songs } = React.useContext(SongsContext);
  let tableData = sData !== undefined ? sData : songs;
  tableData = tableData.length > 0 ? songsTableUtils.addButtons(tableData, token, dispatch) : [];
  // eslint-disable-next-line no-console
  console.log(test);
  return (
    <div style={{ padding: '10px', marginBottom: '-20px' }}>
      {tableData.length > 0 ? (
        <MUIDataTable
          options={{
            filterType: 'dropdown',
            pagination: false,
            responsive: 'standard',
            filter: false,
            download: false,
            search: false,
            print: false,
            viewColumns: false,
            selectableRows: 'none',
            fixedHeader: false,
          }}
          columns={['modify', 'title', 'url', 'artist', 'category', 'album', 'image', 'composer', 'year']}
          data={tableData}
          title=""
        />
      ) : null }
    </div>
  );
};
export default SongsTable;
