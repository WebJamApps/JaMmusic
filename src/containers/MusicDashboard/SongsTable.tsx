import React from 'react';
import MUIDataTable from 'mui-datatables';
import { ISong, SongsContext } from '../../providers/Songs.provider';
import songsTableUtils from './songsTableUtils';

type Props = {
  sData?:ISong[]
};
const SongsTable:React.FC<Props> = ({ sData }): JSX.Element => {
  const { test, songs } = React.useContext(SongsContext);
  let tableData = sData !== undefined ? sData : songs;
  tableData = tableData.length > 0 ? songsTableUtils.addButtons(tableData) : [];
  // eslint-disable-next-line no-console
  console.log(test);
  return (
    <div style={{ padding: '10px', marginBottom: '-20px' }}>
      {tableData.length > 0 ? (
        <MUIDataTable
          options={{
            filterType: 'dropdown',
            pagination: false,
            responsive: 'standard',
            filter: false,
            download: false,
            search: false,
            print: false,
            viewColumns: false,
            selectableRows: 'none',
            fixedHeader: false,
          }}
          columns={['modify', 'title', 'url', 'artist', 'category', 'album', 'image', 'composer', 'year']}
          data={tableData}
          title=""
        />
      ) : null }
    </div>
  );
};
export default SongsTable;
