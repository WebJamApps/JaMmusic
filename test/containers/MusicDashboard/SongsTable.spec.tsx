import React from 'react';
import { shallow } from 'enzyme';
import SongsTable from '../../../src/containers/MusicDashboard/SongsTable';

describe('SongsTable', () => {
  it('is defined', () => {
    expect(SongsTable).toBeDefined();
  });
  it('renders correctly', () => {
    const wrapper = shallow(<SongsTable token="token" dispatch={jest.fn()} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('does not make the table if there are no songs', () => {
    const wrapper = shallow(<SongsTable token="token" sData={[]} dispatch={jest.fn()} />);
    expect(wrapper.find('div').get(0)).toBeDefined();
  });
});
