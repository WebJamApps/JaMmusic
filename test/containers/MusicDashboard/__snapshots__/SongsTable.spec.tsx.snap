// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SongsTable renders correctly 1`] = `ShallowWrapper {}`;
